### ISSUE TITLE
autograd doesn't follow broadcasting

### ISSUE THREAD

**Issue by szagoruyko** [pytorch/pytorch#1787]
Gradient shape is not adjusted to follow broadcasting and autograd fails. Backwards of these two fail, for example:

* ```torch.randn(1) * torch.randn(5,4)```
* ```torch.randn(4,1) * torch.randn(1,4)```

grad/variable shape mismatch:

```python
import torch
from torch.autograd import Variable

# no broadcasting

a = Variable(torch.randn(1), requires_grad=True)
b = Variable(torch.randn(5,4), requires_grad=True)

(a.expand_as(b) * b).sum().backward()

print 'a:', a.size(), a.grad.size()
print 'b:', b.size(), b.grad.size()


# broadcasting failure case

a = Variable(torch.randn(1), requires_grad=True)
b = Variable(torch.randn(5,4), requires_grad=True)

(a * b).sum().backward()

print 'a:', a.size(), a.grad.size() # <-- grad size is different with tensor here
print 'b:', b.size(), b.grad.size()
```

**Comment by gchanan** [comment]
Thanks for the report, I'm looking into it.

**Comment by gchanan** [comment]
The issue is that the Broadcast code gets generated at the cwrap level, which only deals with tensors, so autograd's expand never ends up getting invoked.

It would be nice if this were to happen automatically, but I don't see an easy way to do it; would maybe take something like unifying tensors/Variables or rewriting some of autograd in C.

This issue is conceptually similar to how pointwise operations of different shapes are handled in autograd, e.g.: https://github.com/pytorch/pytorch/blob/5a63a6d47fb20f65546c8d4a9195f04a8ea559c0/torch/autograd/_functions/basic_ops.py#L55 where the backwards has to reshape the output based on the tensor function changing the shape of the inputs.

Probably the easiest way to implement this is just to add expand calls to the relevant functions in variable.py; sort of unfortunate that each function needs to be modified but I don't see an easier way.

**Comment by gchanan** [comment]
Actually, it's probably better to deal with the expansion directly in the autograd Functions; if we do the expansion in variable.py, the tensor functions will never actually do the expansion and we'd need to take some care to deal with error messages, pointwise fallback, etc. correctly.  If we save the sizes of the input tensors and we know the operation (in forward), it should be possible to do the equivalent of Expand.backward in the Functions themselves.

**Comment by gchanan** [comment]
I think this can be closed.
